File: convert-levels.py
#!/usr/bin/env python3

"""
Ball Sort Puzzle Level Converter
Converts your distribution_*.json files to proper game levels
Since all balls are the same color, this is more like a liquid pouring puzzle
"""

import json
import os
import glob
import re
from typing import List, Dict, Any

class BallSortConverter:
    def __init__(self):
        # Since all balls are same color, we'll use a single color
        self.ball_color = "blue"
    
    def convert_puzzle(self, puzzle_data: Dict[str, Any], level_id: int, original_number: int) -> Dict[str, Any]:
        """Convert puzzle data to game format"""
        puzzle = puzzle_data["puzzle"]
        initial_state = puzzle["initialState"]
        bottle_capacities = puzzle["bottleCapacities"]
        desired_level = puzzle["desiredLevel"]
        moves_to_solve = puzzle["movesToSolve"]
        
        # Create tubes based on initial state
        tubes = []
        for i, ball_count in enumerate(initial_state):
            capacity = bottle_capacities[i]
            
            # Create balls array - all same color since that's your format
            balls = [self.ball_color] * ball_count
            
            tubes.append({
                "id": i,
                "balls": balls,
                "capacity": capacity
            })
        
        # Since all balls are the same color, we only have one color
        colors = [self.ball_color] if any(initial_state) else []
        
        return {
            "levelId": level_id,
            "name": f"Level {level_id} - Puzzle {original_number}",
            "difficulty": self.get_difficulty(moves_to_solve, len(tubes), sum(initial_state)),
            "tubes": tubes,
            "colors": colors,
            "moves": 0,
            "minMoves": moves_to_solve,
            "stars": {
                "1": max(moves_to_solve * 3, moves_to_solve + 10),
                "2": max(moves_to_solve * 2, moves_to_solve + 5),
                "3": moves_to_solve
            },
            "originalFile": f"distribution_23555_4_{original_number}_solution.json",
            "puzzleType": "liquid_pouring",  # Different from color sorting
            "desiredLevel": desired_level,
            "solutionSteps": puzzle.get("solutionSteps", [])
        }
    
    def get_difficulty(self, moves: int, tube_count: int, total_balls: int) -> str:
        """Determine difficulty based on various factors"""
        # Factor in moves, tubes, and balls for complexity
        complexity_score = moves + (tube_count * 0.5) + (total_balls * 0.1)
        
        if complexity_score <= 6:
            return "easy"
        elif complexity_score <= 12:
            return "medium"
        else:
            return "hard"
    
    def extract_puzzle_number(self, filename: str) -> int:
        """Extract puzzle number from filename"""
        match = re.search(r'distribution_23555_4_(\d+)_solution\.json', filename)
        return int(match.group(1)) if match else 0

def convert_all_levels():
    """Convert all distribution files to game levels"""
    converter = BallSortConverter()
    levels_dir = "./levels"
    
    # Get all distribution files
    pattern = os.path.join(levels_dir, "distribution_23555_4_*_solution.json")
    files = sorted(glob.glob(pattern), key=lambda x: converter.extract_puzzle_number(os.path.basename(x)))
    
    print(f"Found {len(files)} puzzle files to convert")
    
    converted_levels = []
    level_index = {
        "version": "1.0.0",
        "totalLevels": len(files),
        "levels": [],
        "difficulties": {
            "easy": {"name": "Easy", "color": "#4CAF50", "levels": []},
            "medium": {"name": "Medium", "color": "#FF9800", "levels": []},
            "hard": {"name": "Hard", "color": "#F44336", "levels": []}
        },
        "puzzleType": "liquid_pouring",
        "description": "Liquid pouring puzzle - move balls between tubes with different capacities"
    }
    
    # Convert each file
    for i, file_path in enumerate(files):
        filename = os.path.basename(file_path)
        puzzle_number = converter.extract_puzzle_number(filename)
        level_id = i + 1
        
        try:
            print(f"Converting {filename} -> level-{level_id:03d}.json")
            
            # Read original file
            with open(file_path, 'r') as f:
                puzzle_data = json.load(f)
            
            # Convert to game format
            game_level = converter.convert_puzzle(puzzle_data, level_id, puzzle_number)
            converted_levels.append(game_level)
            
            # Write converted level file
            output_filename = f"level-{level_id:03d}.json"
            output_path = os.path.join(levels_dir, output_filename)
            
            with open(output_path, 'w') as f:
                json.dump(game_level, f, indent=2)
            
            # Add to index
            level_index["levels"].append({
                "id": level_id,
                "file": output_filename,
                "name": game_level["name"],
                "difficulty": game_level["difficulty"],
                "unlocked": level_id == 1,  # Only first level unlocked
                "completed": False,
                "stars": 0,
                "bestMoves": None
            })
            
            # Add to difficulty category
            level_index["difficulties"][game_level["difficulty"]]["levels"].append(level_id)
            
        except Exception as e:
            print(f"Failed to convert {filename}: {str(e)}")
    
    # Write updated level index
    index_path = os.path.join(levels_dir, "levels-index.json")
    with open(index_path, 'w') as f:
        json.dump(level_index, f, indent=2)
    
    print(f"\n✅ Successfully converted {len(converted_levels)} levels!")
    print(f"📊 Difficulty distribution:")
    print(f"   Easy: {len(level_index['difficulties']['easy']['levels'])} levels")
    print(f"   Medium: {len(level_index['difficulties']['medium']['levels'])} levels")
    print(f"   Hard: {len(level_index['difficulties']['hard']['levels'])} levels")
    
    return converted_levels

if __name__ == "__main__":
    convert_all_levels()
    print("\n🎮 Level conversion complete!")

File: convert-levels.js
#!/usr/bin/env node

/**
 * Ball Sort Puzzle Level Converter
 * Converts your distribution_*.json files to proper game levels
 */

const fs = require('fs');
const path = require('path');

class BallSortConverter {
  constructor() {
    this.colors = ['red', 'blue', 'green', 'yellow', 'purple', 'orange', 'pink', 'brown', 'gray', 'cyan'];
  }

  /**
   * Convert puzzle data to game format with proper color distribution
   */
  convertPuzzle(puzzleData, levelId, originalNumber) {
    const { initialState, bottleCapacities, desiredLevel, movesToSolve } = puzzleData.puzzle;
    
    // Calculate how many colors we need based on desiredLevel and total balls
    const totalBalls = initialState.reduce((sum, count) => sum + count, 0);
    const numColors = Math.ceil(totalBalls / desiredLevel);
    const ballsPerColor = desiredLevel;
    
    // Create color distribution
    const allBalls = [];
    for (let i = 0; i < numColors; i++) {
      const color = this.colors[i % this.colors.length];
      for (let j = 0; j < ballsPerColor; j++) {
        allBalls.push(color);
      }
    }
    
    // Shuffle the balls to create a challenging initial state
    this.shuffleArray(allBalls);
    
    // Create tubes
    const tubes = [];
    let ballIndex = 0;
    
    for (let i = 0; i < initialState.length; i++) {
      const ballCount = initialState[i];
      const capacity = bottleCapacities[i];
      const balls = [];
      
      // Fill this tube with balls
      for (let j = 0; j < ballCount; j++) {
        if (ballIndex < allBalls.length) {
          balls.push(allBalls[ballIndex++]);
        }
      }
      
      tubes.push({
        id: i,
        balls: balls,
        capacity: capacity
      });
    }
    
    // Get unique colors used
    const usedColors = [...new Set(allBalls)];
    
    return {
      levelId: levelId,
      name: `Level ${levelId} - Puzzle ${originalNumber}`,
      difficulty: this.getDifficulty(movesToSolve, tubes.length, numColors),
      tubes: tubes,
      colors: usedColors,
      moves: 0,
      minMoves: movesToSolve,
      stars: {
        "1": Math.ceil(movesToSolve * 2.5),
        "2": Math.ceil(movesToSolve * 1.8),
        "3": movesToSolve
      },
      originalFile: `distribution_23555_4_${originalNumber}_solution.json`
    };
  }
  
  /**
   * Determine difficulty based on various factors
   */
  getDifficulty(moves, tubeCount, colorCount) {
    const complexity = moves + (tubeCount * 0.5) + (colorCount * 0.3);
    
    if (complexity <= 8) return "easy";
    if (complexity <= 15) return "medium";
    return "hard";
  }
  
  /**
   * Shuffle array in place (Fisher-Yates algorithm)
   */
  shuffleArray(array) {
    for (let i = array.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [array[i], array[j]] = [array[j], array[i]];
    }
  }
  
  /**
   * Extract puzzle number from filename
   */
  extractPuzzleNumber(filename) {
    const match = filename.match(/distribution_23555_4_(\d+)_solution\.json/);
    return match ? parseInt(match[1]) : null;
  }
}

/**
 * Convert all distribution files to game levels
 */
async function convertAllLevels() {
  const converter = new BallSortConverter();
  const levelsDir = './levels';
  
  try {
    // Get all distribution files
    const files = fs.readdirSync(levelsDir)
      .filter(file => file.startsWith('distribution_') && file.endsWith('_solution.json'))
      .sort((a, b) => {
        const numA = converter.extractPuzzleNumber(a);
        const numB = converter.extractPuzzleNumber(b);
        return numA - numB;
      });
    
    console.log(`Found ${files.length} puzzle files to convert`);
    
    const convertedLevels = [];
    const levelIndex = {
      version: "1.0.0",
      totalLevels: files.length,
      levels: [],
      difficulties: {
        easy: { name: "Easy", color: "#4CAF50", levels: [] },
        medium: { name: "Medium", color: "#FF9800", levels: [] },
        hard: { name: "Hard", color: "#F44336", levels: [] }
      }
    };
    
    // Convert each file
    for (let i = 0; i < files.length; i++) {
      const filename = files[i];
      const puzzleNumber = converter.extractPuzzleNumber(filename);
      const levelId = i + 1;
      
      try {
        console.log(`Converting ${filename} -> level-${levelId.toString().padStart(3, '0')}.json`);
        
        // Read original file
        const filePath = path.join(levelsDir, filename);
        const puzzleData = JSON.parse(fs.readFileSync(filePath, 'utf8'));
        
        // Convert to game format
        const gameLevel = converter.convertPuzzle(puzzleData, levelId, puzzleNumber);
        convertedLevels.push(gameLevel);
        
        // Write converted level file
        const outputFilename = `level-${levelId.toString().padStart(3, '0')}.json`;
        const outputPath = path.join(levelsDir, outputFilename);
        fs.writeFileSync(outputPath, JSON.stringify(gameLevel, null, 2));
        
        // Add to index
        levelIndex.levels.push({
          id: levelId,
          file: outputFilename,
          name: gameLevel.name,
          difficulty: gameLevel.difficulty,
          unlocked: levelId === 1, // Only first level unlocked
          completed: false,
          stars: 0,
          bestMoves: null
        });
        
        // Add to difficulty category
        levelIndex.difficulties[gameLevel.difficulty].levels.push(levelId);
        
      } catch (error) {
        console.error(`Failed to convert ${filename}:`, error.message);
      }
    }
    
    // Write updated level index
    const indexPath = path.join(levelsDir, 'levels-index.json');
    fs.writeFileSync(indexPath, JSON.stringify(levelIndex, null, 2));
    
    console.log(`\n✅ Successfully converted ${convertedLevels.length} levels!`);
    console.log(`📊 Difficulty distribution:`);
    console.log(`   Easy: ${levelIndex.difficulties.easy.levels.length} levels`);
    console.log(`   Medium: ${levelIndex.difficulties.medium.levels.length} levels`);
    console.log(`   Hard: ${levelIndex.difficulties.hard.levels.length} levels`);
    
    return convertedLevels;
    
  } catch (error) {
    console.error('Failed to convert levels:', error);
    return [];
  }
}

// Run conversion if called directly
if (require.main === module) {
  convertAllLevels().then(() => {
    console.log('\n🎮 Level conversion complete!');
  });
}

module.exports = { BallSortConverter, convertAllLevels };
